### PR TITLE
Debugger part 3: display call stack and variable values

### PR DESCRIPTION
**Pull request by jfehrle**
@MSoegtropIMC I hope you'll try this version soon.  Please see the documentation in #14281.  In addition: Clicking on call stack entries (or the Up/Down keys) jump to the associated location and show variable values for that stack frame.  In the variables panel, click on the triangle to see lengthy values.  I recommend detaching the debugger panel for debugging multi-file scripts.

Some current limitations:
- Some variable values are not shown in a useful form, e.g. as "<genarg:tacvalue>"

Please create issues for any bugs you find, including usability issues.  Email me with your questions and comments.  I'll address those ASAP.

~Depends on and incorporates #14252~

Overlay:
- https://github.com/whonore/Coqtail/pull/209

### CODE REVIEW

**Comment by jfehrle** [comment]
@MSoegtropIMC Do you plan to try this sometime soon?  I've made several fixes and improvements, but it looks like I may need to rebase it before GitHub shows the build info on this page.

**Comment by jfehrle** [comment]
@ppedrot Give me a couple of days to fix the broken tests and do some minor cleanup before you dive too deeply into reviewing this.

**Comment by jfehrle** [comment]
@ppedrot Ready for review.

**Comment by Zimmi48** [comment]
If you intend to squash your commit before the PR is merged, then you can safely ignore the linter failure. It is failing because some intermediate commit fails to build (with warnings as errors).

File: doc/changelog/09-coqide/14644-debugger_stack.rst
Comment (Zimmi48) [inline doc/changelog/09-coqide/14644-debugger_stack.rst:2]: No need to repeat "add":
```suggestion
  A preliminary visual debugger.  Supports setting breakpoints
```

**Comment by jfehrle** [comment]
Yes, I expect to squash before merging.

**Review by ejgallego — CHANGES_REQUESTED** [review]
API additions to `DebugHook` look fairly strange, and are undocumented.

I'd comment on them, but without documentation on the API it is hard to understand the intent.

**edit:** after a quick look, some comments:
- `isTerminal` must go away; the debugger interface should be 100% backend agnostic. Current need is due to a deficiency in the protocol.
- duplication between `breakpoints` and `ide_breakpoints` is a no go
- the protocol should be extended to deal with breakpoint setup properly, and the references must be removed; see the debug adapter protocol for examples
- similiarly, the mutable state should be removed and a request for Stack Trace must be implemented
- remove the forward hook and instead provide the reified information in the protocol answer

**Comment by jfehrle** [comment]
Ali, I appreciate your comments and suggestions, but I'd like to keep the PR focused on the code review.  Please put your thoughts in issues rather than here (where they are likely to be lost).

> Can we customise this or at least set the default to something less blue? It's currently very similar to highlighting text.

See Edit / Preferences / Colors.  The setting will be persistent.

> Might it also be a good idea to have a help menu in the debug drop down, especially hinting that you need to do Set Ltac Debug or equivalent.

Not sure what you have in mind.  A link to the manual?  (open a new issue, don't reply here)

**Comment by maximedenes** [comment]
> My request is that refs should not be exposed _in the public API_ in `debugHook`, but I think it is fine and even reasonable to have tactic_debug store the debugger state in a ref [privately]

That's literaly what was said during the Coq call, AFAICT :)

**Comment by jfehrle** [comment]
>> have tactic_debug store the debugger state in a ref [privately]
> That's literaly what was said during the Coq call, AFAICT :)

I don't recall hearing that in the meeting.  In the subsequent discussion here, @ppedrot wanted to put the breakpoints in `TacStore`.  Hopefully he will not object to storing it in a ref in `tactic_debug`.

**Comment by jfehrle** [comment]
> No. The format exported to UI's should be the same regardless of the backend involved. On the contrary, they should not export anything, but you may need both debuggers to communicate at their gateway point, that's not to be exposed on debugHook

OK, we'll perhaps continue discussion on this point in a future PR.

File: toplevel/coqtop.ml
Comment (jfehrle) [inline toplevel/coqtop.ml]: `Vars` has a value of type `(string * Pp.t) list`.  Is there a simple way to marshall this using the existing type definition for the `Db_vars` message in `Xmlprotocol`?  If I need to write new marshalling code, what should that look like?

**Review by jfehrle — COMMENTED** [review]
@ejgallego What's the easiest way to handle the marshalling for Answer?  See the last commit.

File: toplevel/coqtop.ml
Comment (ejgallego) [reply toplevel/coqtop.ml]: Should be just `of_pair of_string of_pp` ? Actually this marshalling code is super annoying, and hindering your work big time.

It wouldn't take long to remove xmlprotocol at all and have `ppx_deriving_yojson` generate the file automatically; tho users of the raw protocol would have to adapt.

File: ide/coqide/idetop.ml
Comment (ejgallego) [reply ide/coqide/idetop.ml:682]: Note you want to put this code in xmlprotocol I think

File: toplevel/coqtop.ml
Comment (jfehrle) [reply toplevel/coqtop.ml]: Close, should be `let of_vars vars = of_list (of_pair of_string of_pp) vars`.  But, as usual, it took a fair bit of time to reverse engineer how to do this.

I want to keep xmlprotocol for this PR, but I'm open to considering changes a bit later.  There are a bunch of things in CoqIDE that would benefit from simplification and refactoring.

Comment (jfehrle) [reply toplevel/coqtop.ml]: However the conversion belongs elsewhere; this particular line of code is for `coqtop`.  This should just be a `failwith`.  I'm not planning to add stack and variable display to the coqtop version.  Indeed, I'd like to drop the support in `coqtop` since it's so much less usable.  But not today.

**Comment by jfehrle** [comment]
@ejgallego Updated.  I changed the set breakpoint, get stack and get vars messages to go through DebugHook.Action and DebugHook.Answer as you asked.  This took significant time; I got stuck a few times.  I hope we're done now except for squashing.

**Comment by jfehrle** [comment]
ping @ejgallego

**Comment by ejgallego** [comment]
> ping @ejgallego

Sorry I missed this due to the terrible github interface and bug (*), will have a look ASAP.

(*) I basically click on the PR updated notification, but I land on the buggy "changes" pages, so I need to get the full notification away :S

**Review by ejgallego — COMMENTED** [review]
Looks much better, thanks @jfehrle ; couple of additional things:
- please document `debugHook.mli` , so I can be sure I understand the protocol
- why do we need to perform `set_break` in idetop.ml ? That deserves a comment too.

**Comment by jfehrle** [comment]
Updated and squashed.  I added the protocol documentation and moved set_break into tactic_debug, although I'll probably need to reverse the latter when we support additional interpreters such as Ltac2.  I believe I fixed the coqtail overlay, so this should be good to go once the tests pass, which I believe they will.

**Comment by jfehrle** [comment]
The two failing jobs appear unrelated.

File: toplevel/coqtop.ml
Comment (ejgallego) [inline toplevel/coqtop.ml:145]: Why is this not supported? You could print it to the terminal? I'd suggest using a proper error message, not an anomaly.

**Review by ejgallego — APPROVED** [review]
Protocol parts look OK to me for now.

Changes in IDE are very extensive tho, including the modification of some tricky stuff regarding GTK and window sizing, etc...

Thus this needs approval of @coq/ltac-maintainers and @coq/coqide-maintainers too.

**Review by ejgallego — CHANGES_REQUESTED** [review]
Oh, `dev/doc/changes.md` need updating [tho I'd prefer to keep the xml protocol changelog in the file itself, and I remember adding it to that effect, but seems it was removed at some point]

**Comment by jfehrle** [comment]
> Thus this needs approval of @coq/ltac-maintainers and @coq/coqide-maintainers too.

@ppedrot would you finish the review and be the assignee?  IIUC, in our Zoom conversation a couple weeks ago you had said the IDE changes were OK.  Let's try to get this completed for 8.15.  Thanks.

**Comment by jfehrle** [comment]
@ejgallego I updated `dev/doc/changes.md`.  Please take a look.

File: toplevel/coqtop.ml
Comment (jfehrle) [reply toplevel/coqtop.ml]: There's currently no code path that can generate these replies in coqtop.  The command line debugger is a candidate for removal--its usability is so much less than the visual debugger that it's probably not worth keeping around at all, let alone enhancing it.

Comment (SkySkimmer) [reply toplevel/coqtop.ml]: you should assert false or anomaly for things that are never supposed to execute, not failwith

Comment (jfehrle) [reply toplevel/coqtop.ml]: What do you have against `failwith`?

I changed it to `anomaly`.  I don't like `assert false` because it doesn't return a message string (unlike `failwith`).  At best you get a file and a line number in a stack trace, which you then need to look up in the exactly matching source code to learn anything, so harder for both developers and users.

**Comment by ejgallego** [comment]
> @ejgallego I updated `dev/doc/changes.md`. Please take a look.

Thanks. The point critical to me is the modification of the `Add` call that breaks compat with virtually all users. Maybe we could be a bit more explicit on what the changes are in the changelog so people know how to adapt their uses?

[There is also a more hacky approach to this kind of changes, which is to introduce a new call `add_with_loc` , so the old call is still preserved, at least for a while, similar to deprecation status]

File: dev/doc/changes.md
Comment (ejgallego) [inline dev/doc/changes.md:14]: I think items 1 and 3 may not be relevant in this file, as they are more like user facing changes? [And already in the main changelog I guess]

I would maybe remove them, and update item 2 to specify the relevant change to add in more detail. Something such as: "before this change, add returned `<xml> .... </xml>` , now, new attributes `<foo> ...` and `<bar>` have been added.

Comment (ejgallego) [reply dev/doc/changes.md:14]: Or provide a pointer to the document where the new return format is spec'ed.

File: ide/coqide/session.ml
Comment (ejgallego) [inline ide/coqide/session.ml:564]: This is the main change in the coqide codebase that worries me, wasn't this submitted as a different PR?

I don't recall the outcome of the discussion, but I'm wary folding this kind of unrelated modifications in this PR.

cc: @coq/coqide-maintainers

Comment (jfehrle) [reply ide/coqide/session.ml:564]: I reversed this.  We earlier decided not to make the change in the PR, #14536.

File: dev/doc/changes.md
Comment (jfehrle) [reply dev/doc/changes.md:14]: Hard to know what should and shouldn't appear in this file.  I removed items 1 and 3.  I added a pointer to `xmlprotocol.md`, which has detailed descriptions of all the messages and responses.  Note that "add" got extra parameters rather than extra return values.

Comment (jfehrle) [reply dev/doc/changes.md:14]: Note that this PR includes an overlay for coqtail to address this very point.  Aside from the PG prototype, are you aware of any other packages that use the xml protocol?

Comment (ejgallego) [reply dev/doc/changes.md:8]: There are some other users indeed, not very actively maintained, but given what we wrote on the protocol file we need to signal the protocol changed [or introduce backwards compatible changes]

Personally I'd keep the changelog for the protocol in the .mli file, in a format `-date-protocol: changes`

**Review by ejgallego — APPROVED** [review]

File: dev/doc/changes.md
Comment (ejgallego) [reply dev/doc/changes.md:8]: Another good practice for this kind of changelogs is to signal explicitly that the change is backwards incompatible, thus you can write:
```
# XML protocol (IMHO better under its own header)
- The add message will now return extra fields for location of the added sentence (warning, backwards-incompatible change, see protocol.md file for more details)
- 4 new "db_*" messages have been added to support the ltac_debugger
```

**Comment by ejgallego** [comment]
> I reversed this. We earlier decided not to make the change in #14536.

Good thanks. @SkySkimmer I added you to the review as not to have a mess with the 8.15 branch; I'd like @ppedrot to give a peek at this if possible.

**Review by ppedrot — APPROVED** [review]
Go ahead.

**Comment by ejgallego** [comment]
@coqbot: merge now

@jfehrle you may consider tweanking the dev/doc/changes.md file if you want.

**Comment by jfehrle** [comment]
@ejgallego @ppedrot Thanks for the review!  Glad we got it into 8.15 just under the wire.

**Comment by mattam82** [comment]
It seems this might have broken VSCoq on master due to the changes in the XML protocol

**Comment by jfehrle** [comment]
Would you open an issue for this and provide more details, e.g.:
- "starting interpretation" meaning trying to execute a tactic or command?
- What are the effects (other than the messages)?
- Were you trying to use the debugger or had a `Set Ltac Debug.` in your script?
- What version of vsCoq/when built?

The debugger changed the parameters and reply for the "add tactic/command" message.
That should be an easy fix in VSCoq if that is the cause.  However, my impression was that
VSCoq was not using the XML protocol.  (8.13 also modified that same message but I never
heard about any issue with VSCoq.)

Attn: VSCoq maintainers @maximedenes @fakusb

**Comment by Blaisorblade** [comment]
> However, my impression was that
> VSCoq was not using the XML protocol.

It certainly is using it — that's how it supports async proofs!

**Comment by jfehrle** [comment]
Please put your comments on the vsCoq issue here: #15282.

**Pull request merged**
